Add indexBuilder.update() method

michaelborn opened the pull request:
This commit adds a few improvements to aid in updating index mappings:

1. A `patch()` method which is far more natural than `.new()` for updating an existing index
2. Removes the warning that the index doesn't exist
3. Added documentation to some method signatures in IndexBuilder.

michaelborn committed:
Add indexBuilder.update() method

This commit adds a few improvements to aid in updating index mappings:

1. An `update()` method which is far more natural than `.new()` for updating an existing index
2. Removes the warning that the index doesn't exist
3. Adds a warning that updating settings on an index is not currently supported. This is an obvious TODO and will hopefully be removed shortly as the capability is added.
4. Added documentation to some method signatures in IndexBuilder.

michaelborn commented:
PS. I tried to run the test suite locally, but there seem to be a few optimizations missing for running tests locally. I received a number of "Read timed out" errors, for instance.

michaelborn committed:
Fix testbox runner URL

This makes it easier to run tests locally (specifically from the CLI.)

michaelborn committed:
Add crude indexBuilder.update() method test

We test that updating a mapping using indexBuilder.update() does not throw an error.

In the future, we should add a way to confirm that the `authorName` field actually exists in the index mapping. At the moment, there isn't a simple way to query the mapping. :/

michaelborn commented:
I added a crude (yet passing) test for the `.update()` method. I don't know of an easy way to query the mapping, so right now the test pretty much only confirms that no error is thrown when calling `.save()`. 🤷

jclausen commented on modules/cbelasticsearch/models/IndexBuilder.cfc:
```diff
@@ -138,6 +145,24 @@ component
+	 * @settings 	{Struct}	Key/value struct of index settings such as `number_of_shards`.
+	 */
+	IndexBuilder function update( string name, any properties, struct settings ) {
+		if ( !isNull( arguments.settings ) && !structIsEmpty( arguments.settings ) ){
```
Some index setting can be updated.  We’re you getting errors?

jclausen replied:
@michaelborn : I need a reply on this one as to why the change here

michaelborn replied:
@jclausen I don't remember off the top of my head why I made this change, but I do think i was getting _some_ sort of error when attempting to update index settings. I was not aware that some settings can be changed on an existing index - perhaps I was trying to update a locked setting, ran into an Elasticsearch error, and promptly decided to add a throw to avoid confusion?
I'll try pulling this down and re-running my code, if I can find where I was working with it. I may have been trying to update the refresh interval on an index during some data imports.

michaelborn commented:
If we can get this properly tested and merged, it resolves #59 . Any help, @jclausen ?

jclausen reviewed (changes requested):
Need to remove the throw on settings update, as some settings may be updated after the index is created.

michaelborn committed:
Merge branch 'development' of github.com:coldbox-modules/cbox-elasticsearch into feature/allow-easy-index-updates

michaelborn committed:
Added test for updating index settings, fixed index mapping update test

michaelborn committed:
Remove throw on changing existing index settings

Per PR feedback.

michaelborn commented:
@jclausen I removed that throw after adding tests to confirm it was totally unnecessary. I also updated the PR for the Hyperclient so we don't get merge conflicts. :smile:

jclausen commented on modules/cbelasticsearch/models/IndexBuilder.cfc:
```diff
@@ -138,6 +145,17 @@ component
+	 * @settings 	{Struct}	Key/value struct of index settings such as `number_of_shards`.
+	 */
+	IndexBuilder function update( string name, any properties, struct settings ) {
+		return this.new( argumentCollection = arguments );
```
I would expect an `update` method to populate and dispatch the save at the same time, since it's a verb. If you don't want to do that, `populate` is a convention we use in a lot of places to hydrate an entity - in which case you could refactor `new` to just call reset and then pass through to `populate` - like cborm's `BaseORMService` does,

michaelborn replied:
Ah, I thought I renamed that method to `patch`?

michaelborn replied:
Odd - this looks like an outdated file. Looks good in `development()`. :+1:

jclausen commented:
@michaelborn : Let's implement a `populate` method that can serve for both `new` and as the populator for `update`.  Then have `update` auto-dispatch the save, since it's a verb.

jclausen commented:
@michaelborn : any chance you can work on this implementation today of the `populate` method? I'm working on a new release and would like to add that.

michaelborn commented:
@jclausen Yep, sure thing!

michaelborn committed:
Merge branch 'development' of github.com:coldbox-modules/cbox-elasticsearch into feature/allow-easy-index-updates

michaelborn committed:
feat: Add indexBuilder.patch() method

Adds an `indexBuilder.patch( "indexName", properties )` convenience method for easily updating mappings or settings on an index.

michaelborn commented:
I pushed up a new `indexBuilder.patch()` method along with a few doc updates. I think this should resolve #59.

My one caveat is my `patch` test is a little poor. I don't know how to query the index and confirm (in the test) that the setting or mapping was added or changed. Any ideas, Jon?

michaelborn commented:
PS. Resolves #59 !
